src/H5Odtype.c

```diff
@@ -452,6 +487,13 @@ H5O__dtype_decode_helper(unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t
                 }
                 if (temp_type->shared->size == 0)
                     HGOTO_ERROR(H5E_DATATYPE, H5E_CANTDECODE, FAIL, "type size can't be zero");
+                if ((dt->shared->u.compnd.memb[dt->shared->u.compnd.nmembs].offset +
```

Comment:
Seems like this should be `dt->shared->u.compnd.nmembs - 1`

Reply:
Actually this one's fine. `dt->shared->u.compnd.nmembs` is the loop index here so this should be valid.